### PR TITLE
[SPARK-4939] move to next locality when no pending tasks

### PR DESCRIPTION
**Pull request by davies**
Currently, if there are different locality in a task set, the tasks with NODE_LOCAL only get scheduled after all the PROCESS_LOCAL tasks are scheduled and timeout with spark.locality.wait.process (3 seconds by default). In local mode, the LocalScheduler will never call resourceOffer() again once it failed to get a task with same locality, then all the NODE_LOCAL tasks will be never scheduled.

This bug could be reproduced by run example python/streaming/stateful_network_wordcount.py, it will hang after finished a batch with some data.

This patch will check whether there is task for current locality level, if not, it will change to next locality level without waiting for `spark.locality.wait.process` seconds. It works for all locality levels.

Because the list of pending tasks are updated lazily, the check can be false-positive, it means it will not move to next locality level even there is no valid pending tasks, it will wait for timeout.

### CODE REVIEW

**Comment by tdas** [comment]
@pwendell Can you please take a look?

**Comment by tdas** [comment]
@kayousterhout Could you take a look at this. This is priority for 1.3 :)

**Comment by pwendell** [comment]
@kayousterhout - can you take a look at this?

**Comment by kayousterhout** [comment]
Why not just periodically revive offers, as we do in the cluster scheduler (see the 3 lines here: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L83)?

**Comment by davies** [comment]
Revive offers periodically can fix the bug, but introduce 1 second delay (by default) for every job, it will be noticeable for streaming workload.

I think it's a bug related to locality, the workaround just hide the bug.

@tdas How do you think?

**Comment by kayousterhout** [comment]
I thought local mode was just used for testing?

I general my preference is to avoid this kind of workaround, which add complexity to the scheduler, make it more brittle, harder to validate that it's working right, harder to maintain, harder to understand, etc. etc.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 1]
```diff
@@ -635,6 +635,28 @@ class TaskSetManagerSuite extends FunSuite with LocalSparkContext with Logging {
     assert(manager.resourceOffer("execA", "host3", NO_PREF).get.index === 2)
   }
 
+  // regression test for SPARK-4939
+  test("node-local tasks should be scheduled right after process-local tasks finished") {
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: Can you just change the name of the test to incorporate the JIRA -- so remove the comment and make it:
test("SPARK-4939: node-local...")

**Comment by kayousterhout** [comment]
Actually I have a slightly different question: should we be doing the same thing for other locality levels?

**Comment by davies** [comment]
@kayousterhout I had updated this PR to support all locality levels.

The timeout of locality depends on that the scheduler backend revive offers periodically, so we should still fix the LocalBackend, or it still get a chance to hang forever (because of the false-positive check in this PR).

**Comment by kayousterhout** [comment]
Is this definitely still an issue with LocalBackend?  It seems like LocalBackend should only ever have PROCESS_LOCAL and NODE_LOCAL tasks.  If it only has NODE_LOCAL (no process local), then everything will work fine.  If it has some PROCESS_LOCAL tasks (the case for this PR), then you don't need to revive offers, because everything gets updated correctly when the last PROCESS_LOCAL task gets scheduled, right?

**Comment by davies** [comment]
@kayousterhout I had added a test case with PROCESS_LOCAL and NO_PREF, it seems that LocalBackend can work well, because the TaskSchedulerImpl will try multiple locality levels in a row, so it will clear the false-positive state before try next level.

So, you are right, LocalBackend can work well with all cases I can image, without revive offers periodically.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 2]
```diff
@@ -495,13 +495,39 @@ private[spark] class TaskSetManager(
...
   private def getAllowedLocalityLevel(curTime: Long): TaskLocality.TaskLocality = {
-    while (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex) &&
-        currentLocalityIndex < myLocalityLevels.length - 1)
-    {
-      // Jump to the next locality level, and remove our waiting time for the current one since
-      // we don't want to count it again on the next one
-      lastLaunchTime += localityWaits(currentLocalityIndex)
-      currentLocalityIndex += 1
+    // remove the emptyList from pendingTasks lazily
+    def hasNonEmptyList(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: In a big cluster, walking through all of these maps each time seems like it could be pretty expensive, right?  Why do this lazily?

Also, it occurred to me that we don't remove tasks from these lists when they're scheduled, so this doesn't completely solve the performance issue (because there could still be a task in one of these lists that will never be scheduled).

This is getting fairly complex to change, and I'm wondering whether it's worthwhile.  Do you have a sense of whether these issues are a problem beyond just the local case where people run streaming jobs that fail?  And if it's only a problem for the local case, should we just fix it in the slightly hacky way by adding a call to revive offers to the local scheduler?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 3]
```diff
@@ -313,7 +313,8 @@ class TaskSetManagerSuite extends FunSuite with LocalSparkContext with Logging {
 
   test("delay scheduling with failed hosts") {
     sc = new SparkContext("local", "test")
-    val sched = new FakeTaskScheduler(sc, ("exec1", "host1"), ("exec2", "host2"))
+    val sched = new FakeTaskScheduler(sc, ("exec1", "host1"), ("exec2", "host2"),
+      ("exec3", "host3"))
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: Why this change?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 2 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: The for-loop will end once there is a list which is not empty, or it will remove all the empty list (not accessed in next time). So I think it will not be expensive even in a big cluster. We also could do it for tasks, then it will fix the false-positive issue.

In short term, this will fix the bug for streaming examples. In long term, this can reduce the latence for streaming (such as updateStateByKey or others needs union).

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala (hunk 3 above)
Comment (davies) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: Because it will change the locality level to NO_PREF if no active executor available.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 2 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: If you have N executors, and they become empty one at a time, then this is a N^2 operation right (because you iterate over all of them -- not just the non-empty ones -- each time)?

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: For the first call, it will be N, then all the execId will be removed from pendingTasksForExecutor, the next call will be 0.

All the empty list will be access once, all the non-empty list, only one of them is accessed.

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I had updated the patch to check whether tasks are scheduled or finished, the above rule also apply to them. It's similar to dequeueTaskFromList()

**Comment by tdas** [comment]
@kayousterhout @davies @pwendell can this be merged for 1.2.1?

**Comment by pwendell** [comment]
@tdas I noted on the JIRA, but since there is collateral effects of this change beyond local mode I think we decided not to put it in 1.2.1.

**Comment by davies** [comment]
@pwendell As pointed out in #4147 , the fix in LocalBackend still have several seconds delay in each stage, it still hurt Spark Streaming. Or should we set `locality.wait` to `0` for all streaming jobs?

**Comment by JoshRosen** [comment]
_(Stream-of-consciousness ahead, mostly for my own benefit as I think through the implications of this PR)_

I'd like to understand whether this patch has any performance implications for Spark jobs in general.  Is there any scenario where this scheduling change might introduce performance regressions?

I guess that a task's locality wait is a sort of hard deadline, where we won't consider scheduling a task at a lower locality level until at least that much time has elapsed.  This sounds like a per-task property, where the deadline for one task shouldn't apply / influence other tasks (e.g. treat each task independently), but the bug reported here sounds like we're applying the locality waits to sets of tasks.

It seems like we always want to attempt to schedule tasks in decreasing order of locality, so if there are unscheduled process-local tasks then we should always attempt to schedule them before any tasks at lower locality levels.  In addition, if there are free slots in the cluster and there are unscheduled tasks in the scheduler's current locality level, then we should schedule those tasks.  If we've exhausted all tasks at a particular locality level, then it makes sense to immediately move onto scheduling at the next lower locality level.

If we have tasks at some high locality level that cannot be scheduled at their preferred locality and there are tasks at a lower locality level that _can_ be scheduled, then I guess we might be concerned about whether scheduling the less-local tasks could rob tasks waiting at a higher locality level of their opportunity to run.  This shouldn't happen, though, since those tasks will already have been offered those resources and turned them down.

Therefore, I think that this patch is a good fix: it doesn't make sense to let a single task with strong preferences to delay / block the scheduling of other tasks with weaker preferences / preferences for other resources.  I'll take a closer look at the code and tests now.

**Comment by JoshRosen** [comment]
> If we have tasks at some high locality level that cannot be scheduled at their preferred locality and there are tasks at a lower locality level that can be scheduled, then I guess we might be concerned about whether scheduling the less-local tasks could rob tasks waiting at a higher locality level of their opportunity to run. This shouldn't happen, though, since those tasks will already have been offered those resources and turned them down.

Actually, on reflection, I guess this gets trickier when you have multiple task sets: in this case, you want to ensure that _all_ unscheduled process-local tasks have turned down a resource before you allow _any_ node-local task to obtain it.

I'm going to try to trace through the interaction between DAGScheduler and TaskSetManager to try to understand how we handle this case.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 4]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
   private def getAllowedLocalityLevel(curTime: Long): TaskLocality.TaskLocality = {
-    while (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex) &&
-        currentLocalityIndex < myLocalityLevels.length - 1)
-    {
-      // Jump to the next locality level, and remove our waiting time for the current one since
-      // we don't want to count it again on the next one
-      lastLaunchTime += localityWaits(currentLocalityIndex)
-      currentLocalityIndex += 1
+    // Remove the scheduled or finished tasks lazily
+    def hasNotScheduledTasks(list: ArrayBuffer[Int]): Boolean = {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Can you give `list` a more descriptive variable name?  Reading top-to-bottom, it's not obvious what it's representing.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 5]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+          list.remove(indexOffset)
+        }
+      }
+      false
+    }
+    // It removes the empty lists after check
+    def hasMoreTasks(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
+      val emptyKeys = new ArrayBuffer[String]
+      val hasTasks = pendingTasks.exists{
+        case (key: String, list: ArrayBuffer[Int]) =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Similarly, could you use a more descriptive name for `key` and `list`?  Is key supposed to be an executor ID?  A host ID?

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: The `key` could be execId or host or rackId, so I change it to id.

**Comment by JoshRosen** [comment]
I took a pass through this, but I'll admit that I'm a little out of my depth w.r.t. scheduler internals here.  I think that I understand the goals / invariants of the scheduler at a high level, but I'm probably not comfortable enough with the internals + perf. optimizations to be comfortable signing off on this.  (Aside: it seems like a lot of the complexity here comes from what seem like (necessary) perf. optimizations, such as not keeping per-task wait timers, using locality levels to index into parallel arrays, etc.).

**Comment by davies** [comment]
For any resource, it will try from the highest locality. TaskSetManager checks the current lowest locality level's wait is non-zero to determine the allowed lowest locality level. If it's zero, then TaskSetManager should assign the possible highest task for the resource, without any waiting. I think it make sense.

There are some logic are perf related, such as the index of task is removed from `list` lazily (only scanned once). Others (using locality levels to index into parallel arrays) are out of this PR's scope.

@rxin Could you also help to look at this PR?

**Comment by rxin** [comment]
cc scheduler czars also... @kayousterhout @markhamstra

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 6]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+      }
+      emptyKeys.foreach(x => pendingTasks.remove(x))
+      hasTasks
+    }
+
+    while (currentLocalityIndex < myLocalityLevels.length - 1) {
+      val moreTasks = myLocalityLevels(currentLocalityIndex) match {
+        case TaskLocality.PROCESS_LOCAL => hasMoreTasks(pendingTasksForExecutor)
+        case TaskLocality.NODE_LOCAL => hasMoreTasks(pendingTasksForHost)
+        case TaskLocality.NO_PREF => pendingTasksWithNoPrefs.isEmpty
```
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: A comment clarifying why the NO_PREF case seems to have the opposite sense of the other cases (not-empty vs. empty) would be helpful.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 7]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+      } else if (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex)) {
+        // Jump to the next locality level, and remove our waiting time for the current one since
+        // we don't want to count it again on the next one
+        lastLaunchTime += localityWaits(currentLocalityIndex)
+        currentLocalityIndex += 1
+        logDebug(s"Move to ${myLocalityLevels(currentLocalityIndex)} after wait for " +
+          s"${localityWaits(currentLocalityIndex)} ms")
+      } else {
+        return myLocalityLevels(currentLocalityIndex)
+      }
```
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: This looks correct to me, but I'm not finding the naming helpful -- in particular, the excess of "index" in `hasNotScheduledTasks`.  I'd suggest something more like this to improve readability (although I'm being a little indecisive about "run" vs. "schedule"):
```scala
    def tasksNeedToBeScheduledFrom(pendingTaskIds: ArrayBuffer[Int]): Boolean = {
      var idOffset = pendingTaskIds.size
      while (idOffset > 0) {
        idOffset -= 1
        val index = pendingTaskIds(idOffset)
        if (copiesRunning(index) == 0 && !successful(index)) {
          return true
        } else {
          pendingTaskIds.remove(idOffset)
        }
      }
      false
    }
    // It removes the empty lists after check
    def noMoreTasksToRunIn(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
      val emptyKeys = new ArrayBuffer[String]
      val moreTasksToRun = pendingTasks.exists {
        case (id: String, pendingTaskIds: ArrayBuffer[Int]) =>
          if (tasksNeedToBeScheduledFrom(pendingTaskIds)) {
            true
          } else {
            emptyKeys += id
            false
          }
      }
      emptyKeys.foreach(x => pendingTasks.remove(x))
      !moreTasksToRun
    }

    while (currentLocalityIndex < myLocalityLevels.length - 1) {
      val noMoreTasksToRun = myLocalityLevels(currentLocalityIndex) match {
        case TaskLocality.PROCESS_LOCAL => noMoreTasksToRunIn(pendingTasksForExecutor)
        case TaskLocality.NODE_LOCAL => noMoreTasksToRunIn(pendingTasksForHost)
        case TaskLocality.NO_PREF => pendingTasksWithNoPrefs.nonEmpty
        case TaskLocality.RACK_LOCAL => noMoreTasksToRunIn(pendingTasksForRack)
      }
      if (noMoreTasksToRun) {
.
.
.
```
`getAllowedLocalityLevel` also doesn't give adequate notice of the side effect that modifies `emptyKeys`, but I haven't got a good solution for that -- other than hoisting and intensifying the comment before `hasMoreTasks`.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 8]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+        case TaskLocality.PROCESS_LOCAL => hasMoreTasks(pendingTasksForExecutor)
+        case TaskLocality.NODE_LOCAL => hasMoreTasks(pendingTasksForHost)
+        case TaskLocality.NO_PREF => pendingTasksWithNoPrefs.isEmpty
+        case TaskLocality.RACK_LOCAL => hasMoreTasks(pendingTasksForRack)
+      }
+      if (!moreTasks) {
+        // Move to next locality level if there is no task for current level
+        lastLaunchTime = curTime
+        logDebug(s"No tasks for locality level ${myLocalityLevels(currentLocalityIndex)} " +
+          s"move to ${myLocalityLevels(currentLocalityIndex + 1)}")
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: nit: "move to..." ==> "so moving to locality level ..."

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 9]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+        currentLocalityIndex += 1
+      } else if (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex)) {
+        // Jump to the next locality level, and remove our waiting time for the current one since
+        // we don't want to count it again on the next one
+        lastLaunchTime += localityWaits(currentLocalityIndex)
+        currentLocalityIndex += 1
+        logDebug(s"Move to ${myLocalityLevels(currentLocalityIndex)} after wait for " +
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: s"Moving to ${..} after waiting for ${..}ms"

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: +1 (is this a mistake?)

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 10]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+      }
+      if (!moreTasks) {
+        // Move to next locality level if there is no task for current level
+        lastLaunchTime = curTime
+        logDebug(s"No tasks for locality level ${myLocalityLevels(currentLocalityIndex)} " +
+          s"move to ${myLocalityLevels(currentLocalityIndex + 1)}")
+        currentLocalityIndex += 1
+      } else if (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex)) {
+        // Jump to the next locality level, and remove our waiting time for the current one since
+        // we don't want to count it again on the next one
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I know this existed before your change, but can you change this comment to: "Jump to the next locality level, and reset lastLaunchTime so that the next locality wait timer doesn't immediately expire"? I think that would make this easier to understand

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: The tasks with NO_PREF should not be in other lists, so do not need to remove them lazily, just check the whole list is empty or not.

I will add a comment.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 7 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: It would also help to write a slightly longer / more informative comment in place of "It removes.." and "Remove the scheduled...".  What about adding the following before noMoreTasksToRun:

"Walk through the list of tasks that can be scheduled at each location and returns true if there are any tasks that still need to be scheduled.  Lazily cleans up tasks that have already been scheduled."

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 11]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+
+    while (currentLocalityIndex < myLocalityLevels.length - 1) {
+      val moreTasks = myLocalityLevels(currentLocalityIndex) match {
+        case TaskLocality.PROCESS_LOCAL => hasMoreTasks(pendingTasksForExecutor)
+        case TaskLocality.NODE_LOCAL => hasMoreTasks(pendingTasksForHost)
+        case TaskLocality.NO_PREF => pendingTasksWithNoPrefs.isEmpty
+        case TaskLocality.RACK_LOCAL => hasMoreTasks(pendingTasksForRack)
+      }
+      if (!moreTasks) {
+        // Move to next locality level if there is no task for current level
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Add something like: "This is a performance optimization: if there are no more tasks that can be scheduled at a particular locality level, there is no point in waiting for the locality wait timeout (SPARK-4939)."

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: But shouldn't it be !pendingTasksWithNoPrefs.isEmpty?

**Comment by kayousterhout** [comment]
Echoing the comments from @markhamstra and @JoshRosen, this change looks correct but is very subtle, so it would be great to improve the commenting throughout to avoid confusion for others who read this code later.  I suggested a few comment improvements; if you have time to make these soon, that would be great since I know you wanted this in for 1.3 and I know the merge window for that is closing rapidly.  Thanks @davies!

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Oh, yes, it's really a mistake! will fix it.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 12]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+      val hasTasks = pendingTasks.exists{
+        case (id: String, tasks: ArrayBuffer[Int]) =>
+          if (hasNotScheduledTasks(tasks)) {
+            true
+          } else {
+            emptyKeys += id
+            false
+          }
+      }
+      emptyKeys.foreach(x => pendingTasks.remove(x))
```
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]:
```scala
emptyKeys.foreach(executorId => pendingTasks.remove(executorId))
```

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 13]
```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
...
+    }
+    // It removes the empty lists after check
+    def hasMoreTasks(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
+      val emptyKeys = new ArrayBuffer[String]
+      val hasTasks = pendingTasks.exists{
+        case (id: String, tasks: ArrayBuffer[Int]) =>
+          if (hasNotScheduledTasks(tasks)) {
+            true
+          } else {
+            emptyKeys += id
```
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: id ==> executorId

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: pendingTasksWithNoPrefs.nonEmpty

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 13 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: The `id` could be executorId, host or rackId.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 12 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Also, the key could be executorId, host, or rackId

Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Ah yes, I see now.  So, leave it as `id` and slip in the `key could be executorId, host, or rackId` comment.

**Comment by davies** [comment]
@kayousterhout @markhamstra @JoshRosen I should had addressed all your comments, they are very helpful, thanks you all!

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 14]
```diff
@@ -506,13 +506,63 @@ private[spark] class TaskSetManager(
...
+      }
+      emptyKeys.foreach(x => pendingTasks.remove(x))
+      hasTasks
+    }
+
+    while (currentLocalityIndex < myLocalityLevels.length - 1) {
+      val moreTasks = myLocalityLevels(currentLocalityIndex) match {
+        case TaskLocality.PROCESS_LOCAL => noMoreTasksToRunIn(pendingTasksForExecutor)
+        case TaskLocality.NODE_LOCAL => noMoreTasksToRunIn(pendingTasksForHost)
+        case TaskLocality.NO_PREF => pendingTasksWithNoPrefs.nonEmpty
```
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Wait, are we going with moreTasks or noMoreTasks?  I think this needs to be `val noMoreTasks`, `case TaskLocality.NO_PREF => pendingTasksWithNoPrefs.isEmpty` and `if (noMoreTasks)` a couple lines down -- or consistently reverse the boolean sense of all that.

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: noMoreTasksToRunIn => moreTasksToRunIn()

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: oops, I just rush it too fast without check it carefully.

**Comment by kayousterhout** [comment]
LGTM (but will wait to merge until @markhamstra and @JoshRosen give the thumb's up, just in case they have any remaining comments)

**Comment by markhamstra** [comment]
LGTMT

**Comment by JoshRosen** [comment]
I understood the high-level idea here and was just waiting for @kayousterhout and @markhamstra to review the details, so this seems fine to me.  Thanks for reviewing this!

**Comment by kayousterhout** [comment]
Thanks all! I've merged this into 1.3 and master.
